### ISSUE TITLE
[JENKINS-51700] Cannot run command on other agent/master when global agent is kubernetes

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/kubernetes-plugin#2059]
<p>Simplified version of my declarative pipeline looks like this:</p>

<pre>
pipeline {
  agent {
    kubernetes {
      cloud <span class="code-quote">'kube-cloud'</span>
      label <span class="code-quote">'kube'</span>
      yaml """
apiVersion: v1
kind: Pod
metadata:
  name: kube
spec:
  containers:
  - name: jnlp
    image: jenkins/jnlp-slave
"""
    }
  }

  stages {
    stage(<span class="code-quote">'step-in-kubernetes'</span>) {
      steps {
        sh <span class="code-quote">'echo inKubernetes'</span>
      }
    }

    stage(<span class="code-quote">'step-in-agent'</span>) {
      agent {
        label <span class="code-quote">'agentX'</span>
      }
      steps {
        sh <span class="code-quote">'echo inAgent'</span>
      }
    }
  }
}
</pre>
<p>In this case pipeline fails on stage 'step-in-agent' as following:</p>

<pre>
Running in Durability level: MAX_SURVIVABILITY
[Pipeline] podTemplate
[Pipeline] {
[Pipeline] node
Still waiting to schedule task
All nodes of label ‘kube’ are offline
Agent kube-3cqmb-dpw7t is provisioned from template Kubernetes Pod Template
Agent specification [Kubernetes Pod Template] (kube):

Running on kube-3cqmb-dpw7t in /home/jenkins/workspace/pipeline-tests
[Pipeline] {
[Pipeline] container
[Pipeline] {
[Pipeline] stage
[Pipeline] { (step-in-kubernetes)
[Pipeline] sh
[pipeline-tests] Running shell script
+ echo inKubernetes
inKubernetes
[Pipeline] }
[Pipeline] <span class="code-comment">// stage
</span>[Pipeline] stage
[Pipeline] { (step-in-agent)
[Pipeline] node
Running on Jenkins in /<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace
[Pipeline] {
[Pipeline] sh
[workspace] Running shell script
/bin/sh: 1: cd: can't cd to /<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace
sh: 1: cannot create /<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace@​tmp/durable-0bd20381/jenkins-log.txt: Directory nonexistent
sh: 1: cannot create /<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace@​tmp/durable-0bd20381/jenkins-result.txt.tmp: Directory nonexistent
mv: cannot stat <span class="code-quote">'/<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace@​tmp/durable-0bd20381/jenkins-result.txt.tmp'</span>: No such file or directory
EXITCODE   0process apparently never started in /<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace@​tmp/durable-0bd20381
[Pipeline] }
[Pipeline] <span class="code-comment">// node
</span>[Pipeline] }
[Pipeline] <span class="code-comment">// stage
</span>[Pipeline] }
[Pipeline] <span class="code-comment">// container
</span>[Pipeline] }
[Pipeline] <span class="code-comment">// node
</span>[Pipeline] }
[Pipeline] <span class="code-comment">// podTemplate
</span>[Pipeline] End of Pipeline
ERROR: script returned exit code -2
Finished: FAILURE</pre>
<p>In my real pipeline, I must run a few steps in parallel on kubernetes containers, and after this I must run some steps on other agent which cannot be in kubernetes container. I know that it is possible to do something like this in scripted pipeline, but it is possible to achieve this in declarative pipeline ?</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mjanczuk.png" title="mjanczuk's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=mjanczuk">mjanczuk</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-51700" target="_blank">Cannot run command on other agent/master when global agent is kubernetes</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: kubernetes-plugin, pipeline-model-definition-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2021-01-05T15:08:13+00:00
<li><b>votes</b>: 9
<li><b>watchers</b>: 12
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Simplified version of my declarative pipeline looks like this:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
  agent {
    kubernetes {
      cloud <span class="code-quote">'kube-cloud'</span>
      label <span class="code-quote">'kube'</span>
      yaml """
apiVersion: v1
kind: Pod
metadata:
  name: kube
spec:
  containers:
  - name: jnlp
    image: jenkins/jnlp-slave
"""
    }
  }

  stages {
    stage(<span class="code-quote">'step-in-kubernetes'</span>) {
      steps {
        sh <span class="code-quote">'echo inKubernetes'</span>
      }
    }

    stage(<span class="code-quote">'step-in-agent'</span>) {
      agent {
        label <span class="code-quote">'agentX'</span>
      }
      steps {
        sh <span class="code-quote">'echo inAgent'</span>
      }
    }
  }
}
</pre>
</div></div>
<p>In this case pipeline fails on stage 'step-in-agent' as following:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Running in Durability level: MAX_SURVIVABILITY
[Pipeline] podTemplate
[Pipeline] {
[Pipeline] node
Still waiting to schedule task
All nodes of label ‘kube’ are offline
Agent kube-3cqmb-dpw7t is provisioned from template Kubernetes Pod Template
Agent specification [Kubernetes Pod Template] (kube):

Running on kube-3cqmb-dpw7t in /home/jenkins/workspace/pipeline-tests
[Pipeline] {
[Pipeline] container
[Pipeline] {
[Pipeline] stage
[Pipeline] { (step-in-kubernetes)
[Pipeline] sh
[pipeline-tests] Running shell script
+ echo inKubernetes
inKubernetes
[Pipeline] }
[Pipeline] <span class="code-comment">// stage
</span>[Pipeline] stage
[Pipeline] { (step-in-agent)
[Pipeline] node
Running on Jenkins in /<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace
[Pipeline] {
[Pipeline] sh
[workspace] Running shell script
/bin/sh: 1: cd: can't cd to /<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace
sh: 1: cannot create /<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace@tmp/durable-0bd20381/jenkins-log.txt: Directory nonexistent
sh: 1: cannot create /<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace@tmp/durable-0bd20381/jenkins-result.txt.tmp: Directory nonexistent
mv: cannot stat <span class="code-quote">'/<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace@tmp/durable-0bd20381/jenkins-result.txt.tmp'</span>: No such file or directory
EXITCODE   0process apparently never started in /<span class="code-keyword">var</span>/lib/jenkins/jobs/pipeline-tests/workspace@tmp/durable-0bd20381
[Pipeline] }
[Pipeline] <span class="code-comment">// node
</span>[Pipeline] }
[Pipeline] <span class="code-comment">// stage
</span>[Pipeline] }
[Pipeline] <span class="code-comment">// container
</span>[Pipeline] }
[Pipeline] <span class="code-comment">// node
</span>[Pipeline] }
[Pipeline] <span class="code-comment">// podTemplate
</span>[Pipeline] End of Pipeline
ERROR: script returned exit code -2
Finished: FAILURE</pre>
</div></div>
<p>In my real pipeline, I must run a few steps in parallel on kubernetes containers, and after this I must run some steps on other agent which cannot be in kubernetes container. I know that it is possible to do something like this in scripted pipeline, but it is possible to achieve this in declarative pipeline ?</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Jenkins ver. 2.107.3 with kubernetes-plugin</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-51700?focusedId=345737&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-345737">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I think this is a combination of a limitation in the kubernetes plugin (you can't run a <tt>node</tt> block inside a <tt>container</tt> block) and a limitation in Declarative (per-stage agents run within the top-level agent block). You may want to investigate sequential stages (<a href="https://jenkins.io/blog/2018/07/02/whats-new-declarative-piepline-13x-sequential-stages/" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/blog/2018/07/02/whats-new-declarative-piepline-13x-sequential-stages/</a>) - you could have a top level <tt>agent none</tt>, then a parent <tt>stage</tt> with the kubernetes agent and a <tt>stages</tt> block of all the stages that need to run on that kubernetes agent, and finally, outside of that parent <tt>stage</tt>, the stage that needs to run on a different agent.</p>
</pre>
</details>

</ul>
</details>

<p>I think this is a combination of a limitation in the kubernetes plugin (you can't run a <tt>node</tt> block inside a <tt>container</tt> block) and a limitation in Declarative (per-stage agents run within the top-level agent block). You may want to investigate sequential stages (<a href="https://jenkins.io/blog/2018/07/02/whats-new-declarative-piepline-13x-sequential-stages/" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/blog/2018/07/02/whats-new-declarative-piepline-13x-sequential-stages/</a>) - you could have a top level <tt>agent none</tt>, then a parent <tt>stage</tt> with the kubernetes agent and a <tt>stages</tt> block of all the stages that need to run on that kubernetes agent, and finally, outside of that parent <tt>stage</tt>, the stage that needs to run on a different agent.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/c9s.svg" title="c9s's avatar" /> c9s:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-51700?focusedId=356108&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-356108">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>same here. I wanted to have a default k8s agent, and in the sub-stage I want to allocate a node for isolating the tests. However, it shows "can not cd into {workspace}" error.</p>
</pre>
</details>

</ul>
</details>

<p>same here. I wanted to have a default k8s agent, and in the sub-stage I want to allocate a node for isolating the tests. However, it shows "can not cd into {workspace}" error.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/stevenschlansker.jpg" title="stevenschlansker's avatar" /> stevenschlansker:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-51700?focusedId=373548&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-373548">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We are also suffering this issue on jenkins 2.176.2 with kubernetes plugin 1.18.1</p>
</pre>
</details>

</ul>
</details>

<p>We are also suffering this issue on jenkins 2.176.2 with kubernetes plugin 1.18.1</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/vlatombe.png" title="vlatombe's avatar" /> vlatombe:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-51700?focusedId=402938&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-402938">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I don't think it can be fixed by the plugin. It is just a syntax limitation which can't be overcome.</p>
</pre>
</details>

</ul>
</details>

<p>I don't think it can be fixed by the plugin. It is just a syntax limitation which can't be overcome.</p>
